Add union type overloads where applicable

## 🚀 Feature Request

Where the API provides overloads with different parameter types, provide variants that accept unions of those parameter types as well.

## Motivation

Let's say we are writing a `head` operation on Node.js readable streams. We want to generate a scenario with an arbitrary stream and number of items to include:

```typescript
interface HeadConstraints {
  upstream: TestStreamConstraints,  // defined elsewhere
  n: number | NatConstraints,
}
```

We want to render this into parameters to pass during our test:
```typescript
function headScenario(constraints: HeadConstraints): Arbitrary<[stream.Readable, number]> {
  return fc.tuple(
    testStream(constraints.upstream),
    nat(constraints.n),  // Error!
  );
}
```

fast-check gives the following error:
```typescript
    src/lib/stream.spec.ts:80:18 - error TS2769: No overload matches this call.
      Overload 1 of 3, '(max: number): Arbitrary<number>', gave the following error.
        Argument of type 'number | NatConstraints' is not assignable to parameter of type 'number'.
          Type 'NatConstraints' is not assignable to type 'number'.
      Overload 2 of 3, '(constraints: NatConstraints): Arbitrary<number>', gave the following error.
        Argument of type 'number | NatConstraints' is not assignable to parameter of type 'NatConstraints'.
          Type 'number' has no properties in common with type 'NatConstraints'.

    80     nat = fc.nat(constraints.n);
```

- `nat` can accept a `number`, but `number | NatConstraints` is not assignable to `number`.
- `nat` can accept a `NatConstraints`, but `number | NatConstraints` is not assignable to `NatConstraints`.

At that point, the compiler gives up.

Of course, `nat` works perfectly well with a argument of type `number | NatConstraints`, but TypeScript cannot infer this automatically (discussion [here](https://github.com/microsoft/TypeScript/issues/14107)). The current solution is to narrow the argument to one type or the other:

```typescript
let nat: Arbitrary<number>;
if (typeof constraints.n === 'number') {
  nat = fc.nat(constraints.n);
} else {
  nat = fc.nat(constraints.n);
}
```

A bit awkward.

The most convenent solution is to provide an overload for `fc.nat`:
```typescript
declare function nat(): Arbitrary<number>;
declare function nat(max: number): Arbitrary<number>;
declare function nat(constraints: NatConstraints): Arbitrary<number>;
declare function nat(constraints: Number | NatConstraints): Arbitrary<number>;  // fixed!
```

`bigInt` would be another candidate for this treatment, and possibly others.

--- a/.changeset/light-eggs-cut.md
+++ b/.changeset/light-eggs-cut.md
@@ -1,0 +1,5 @@
+---
+'fast-check': patch
+---
+
+Add union type overloads for `nat` and `bigInt`

--- a/packages/fast-check/src/arbitrary/bigInt.ts
+++ b/packages/fast-check/src/arbitrary/bigInt.ts
@@ -81,6 +81,15 @@ function bigInt(min: bigint, max: bigint): Arbitrary<bigint>;
  * @public
  */
 function bigInt(constraints: BigIntConstraints): Arbitrary<bigint>;
+/**
+ * For bigint between min (included) and max (included)
+ *
+ * @param args - Either min/max bounds as an object or constraints to apply when building instances
+ *
+ * @remarks Since 2.6.0
+ * @public
+ */
+function bigInt(...args: [] | [bigint, bigint] | [BigIntConstraints]): Arbitrary<bigint>;
 function bigInt(...args: [] | [bigint, bigint] | [BigIntConstraints]): Arbitrary<bigint> {
   const constraints = buildCompleteBigIntConstraints(extractBigIntConstraints(args));
   if (constraints.min > constraints.max) {

--- a/packages/fast-check/src/arbitrary/nat.ts
+++ b/packages/fast-check/src/arbitrary/nat.ts
@@ -42,6 +42,15 @@ function nat(max: number): Arbitrary<number>;
  * @public
  */
 function nat(constraints: NatConstraints): Arbitrary<number>;
+/**
+ * For positive integers between 0 (included) and max (included)
+ *
+ * @param arg - Either a maximum number or constraints to apply when building instances
+ *
+ * @remarks Since 2.6.0
+ * @public
+ */
+function nat(arg?: number | NatConstraints): Arbitrary<number>;
 function nat(arg?: number | NatConstraints): Arbitrary<number> {
   const max = typeof arg === 'number' ? arg : arg && arg.max !== undefined ? arg.max : 0x7fffffff;
   if (max < 0) {

--- a/packages/fast-check/test/unit/arbitrary/bigInt.spec.ts
+++ b/packages/fast-check/test/unit/arbitrary/bigInt.spec.ts
@@ -1,6 +1,6 @@
 import { describe, it, expect, vi } from 'vitest';
 import * as fc from 'fast-check';
-import { bigInt } from '../../../src/arbitrary/bigInt';
+import { bigInt, type BigIntConstraints } from '../../../src/arbitrary/bigInt';
 
 import { fakeArbitrary } from './__test-helpers__/ArbitraryHelpers';
 import { declareCleaningHooksForSpies } from './__test-helpers__/SpyCleaner';
@@ -88,5 +88,45 @@ describe('bigInt', () => {
         // Act / Assert
         expect(() => bigInt({ min: high, max: low })).toThrowError();
       }),
+    ));
+
+  it('should handle union type of [number, number] | BigIntConstraints', () =>
+    fc.assert(
+      fc.property(
+        fc.oneof(
+          // no argument
+          fc.tuple(),
+          // min, max range
+          fc.tuple(fc.bigInt(), fc.bigInt()).map<[bigint, bigint]>((t) => (t[0] < t[1] ? [t[0], t[1]] : [t[1], t[0]])),
+          // both min and max defined
+          fc.tuple(
+            fc.record({ min: fc.bigInt(), max: fc.bigInt() }).map((c) => ({
+              min: c.min < c.max ? c.min : c.max,
+              max: c.min < c.max ? c.max : c.min,
+            })),
+          ),
+          // one or the other maybe defined
+          fc.tuple(fc.record({ min: fc.option(fc.bigInt(), { nil: undefined }) })),
+          fc.tuple(fc.record({ max: fc.option(fc.bigInt(), { nil: undefined }) })),
+        ),
+        (args: [] | [bigint, bigint] | [BigIntConstraints]) => {
+          // Arrange
+          const [expectedMin, expectedMax] =
+            args.length === 0 ? [null, null] : args.length === 1 ? [args[0].min, args[0].max] : [args[0], args[1]];
+          const instance = fakeBigIntArbitrary();
+          const BigIntArbitrary = vi.spyOn(BigIntArbitraryMock, 'BigIntArbitrary');
+          BigIntArbitrary.mockImplementation(() => instance);
+
+          // Act
+          const arb = bigInt(...args);
+
+          // Assert
+          expect(BigIntArbitrary).toHaveBeenCalledWith(
+            expectedMin ?? expect.any(BigInt),
+            expectedMax ?? expect.any(BigInt),
+          );
+          expect(arb).toBe(instance);
+        },
+      ),
     ));
 });

--- a/packages/fast-check/test/unit/arbitrary/double.spec.ts
+++ b/packages/fast-check/test/unit/arbitrary/double.spec.ts
@@ -190,7 +190,16 @@ describe('double', () => {
           expect(bigInt).toHaveBeenCalledTimes(2);
           const constraintsNoNaN = bigInt.mock.calls[0];
           const constraintsWithNaN = bigInt.mock.calls[1];
+          if (constraintsNoNaN.length !== 1) {
+            expect.fail(`Expected bigInt to be called with one constraints object; got ${constraintsNoNaN}`);
+          }
+          if (constraintsWithNaN.length !== 1) {
+            expect.fail(`Expected bigInt to be called with one constraints object; got ${constraintsWithNaN}`);
+          }
           if (max > Number.MIN_VALUE || (max > 0 && !ct.maxExcluded)) {
+            if (constraintsNoNaN.length !== 1) {
+              expect.fail(`Expected bigInt to be called with one constraints object; got ${constraintsNoNaN}`);
+            }
             // max > 0  --> NaN will be added as the greatest value
             expect(constraintsWithNaN[0]).toEqual({
               min: constraintsNoNaN[0].min,
@@ -226,6 +235,12 @@ describe('double', () => {
             double({ ...ct, noNaN: true });
             const arb = double(ct);
             // Extract NaN "index"
+            if (bigInt.mock.calls[0].length !== 1) {
+              expect.fail(`Expected bigInt to be called with one constraints object; got ${bigInt.mock.calls[0]}`);
+            }
+            if (bigInt.mock.calls[1].length !== 1) {
+              expect.fail(`Expected bigInt to be called with one constraints object; got ${bigInt.mock.calls[1]}`);
+            }
             const [{ min: minNonNaN }] = bigInt.mock.calls[0];
             const [{ min: minNaN, max: maxNaN }] = bigInt.mock.calls[1];
             const indexForNaN = minNonNaN !== minNaN ? minNaN : maxNaN;

--- a/packages/fast-check/test/unit/arbitrary/nat.spec.ts
+++ b/packages/fast-check/test/unit/arbitrary/nat.spec.ts
@@ -1,6 +1,6 @@
 import { describe, it, expect, vi } from 'vitest';
 import * as fc from 'fast-check';
-import { nat } from '../../../src/arbitrary/nat';
+import { nat, type NatConstraints } from '../../../src/arbitrary/nat';
 
 import { fakeArbitrary } from './__test-helpers__/ArbitraryHelpers';
 
@@ -96,4 +96,30 @@ describe('nat', () => {
       }),
     );
   });
+
+  it('should handle union type of number | NatConstraints', () =>
+    fc.assert(
+      fc.property(
+        fc.oneof(
+          fc.constant(undefined),
+          fc.integer({ min: 0 }),
+          fc.record({}),
+          fc.record({ max: fc.integer({ min: 0 }) }),
+        ),
+        (constraints: number | NatConstraints | undefined) => {
+          // Arrange
+          const expectedMax = typeof constraints === 'number' ? constraints : (constraints?.max ?? null);
+          const instance = fakeIntegerArbitrary();
+          const IntegerArbitrary = vi.spyOn(IntegerArbitraryMock, 'IntegerArbitrary');
+          IntegerArbitrary.mockImplementation(() => instance);
+
+          // Act
+          const arb = constraints === undefined ? nat() : nat(constraints);
+
+          // Assert
+          expect(IntegerArbitrary).toHaveBeenCalledWith(0, expectedMax ?? expect.anything());
+          expect(arb).toBe(instance);
+        },
+      ),
+    ));
 });

